[SeparateConstOffsetFromGEP] Decompose constant xor operand if possible

sgundapa opened the pull request:
Try to transform XOR(A, B+C) in to XOR(A,C) + B where XOR(A,C) becomes
the base for memory operations. This transformation is true under the
following conditions
Check 1 -  B and C are disjoint.
Check 2 - XOR(A,C) and B are disjoint.

This transformation is beneficial particularly for GEPs because
Disjoint OR operations often map better to addressing modes than XOR.
This can enable further optimizations in the GEP offset folding pipeline

sgundapa committed:
Decompose Xors that are fed to GEPs

NOTE: This patch is not to be merged, just for evaluation.

sgundapa commented on llvm/lib/Transforms/Scalar/SeparateConstOffsetFromGEP.cpp:
```diff
@@ -1162,6 +1168,162 @@ bool SeparateConstOffsetFromGEP::splitGEP(GetElementPtrInst *GEP) {
+
+  auto UserInst = findClosestSequentialXor(A, I);
+
+  BinaryOperator *UserBO = cast<BinaryOperator>(UserInst);
```
I will add a nullptr check here before I use this instruction.

sgundapa commented on llvm/lib/Transforms/Scalar/SeparateConstOffsetFromGEP.cpp:
```diff
@@ -1162,6 +1168,162 @@ bool SeparateConstOffsetFromGEP::splitGEP(GetElementPtrInst *GEP) {
+
+  // Check Disjointness A & C2 == 0.
+  KnownBits KnownA(BitWidth);
+  AssumptionCache *AC = nullptr;
```
I need to clean this up

sgundapa commented on llvm/lib/Transforms/Scalar/SeparateConstOffsetFromGEP.cpp:
```diff
@@ -1162,6 +1168,162 @@ bool SeparateConstOffsetFromGEP::splitGEP(GetElementPtrInst *GEP) {
+#if 0
+  // Preserve metadata
+  if (Instruction *NewOrInst = dyn_cast<Instruction>(NewOr)) {
+    NewOrInst->copyMetadata(I);
```
will add this to preserve the metadata

arsenm replied:
There could be more metadata than the debug lines, but it's not necessarily trivially correct to preserve it all

jrbyrnes commented on llvm/lib/Transforms/Scalar/SeparateConstOffsetFromGEP.cpp:
```diff
@@ -1162,6 +1168,162 @@ bool SeparateConstOffsetFromGEP::splitGEP(GetElementPtrInst *GEP) {
+  Instruction *FoundUserInst = nullptr; // Instruction Y
+  APInt C0_APInt;
+
+  auto UserInst = findClosestSequentialXor(A, I);
```
It seems to me that we should still decompose the xor (if it can be made disjoint) -- even if there is no preexisting compatible xor

sgundapa replied:
I will try this as a follow up patch to this. Let me know what you think

sgundapa committed:
Push correctness fixes

sgundapa committed:
Update lit test

jmmartinez commented on llvm/lib/Transforms/Scalar/SeparateConstOffsetFromGEP.cpp:
```diff
@@ -1162,6 +1167,179 @@ bool SeparateConstOffsetFromGEP::splitGEP(GetElementPtrInst *GEP) {
+        ClosestUser = UserInst;
+      else {
+        // Compare instruction positions.
+        if (UserInst->comesBefore(ClosestUser)) {
```
This would assert if the instructions are not in the same parent block, right ?

sgundapa replied:
The naming is confusing and probably wrong. I will change the name of this function.

B = xor (A, 32)
C = xor (A, 2080)  --> modified to C = XOR(B, 2048)
D = xor( A, 4128) --> modified to D = XOR(B, 4096) , here, when I scan for XOR, I see both B and C. This function tries to get B.

jmmartinez commented on llvm/lib/Transforms/Scalar/SeparateConstOffsetFromGEP.cpp:
```diff
@@ -1162,6 +1167,179 @@ bool SeparateConstOffsetFromGEP::splitGEP(GetElementPtrInst *GEP) {
+  }
+
+  IRBuilder<> Builder(&I);
+  Builder.SetInsertPoint(&I);
```
The insert point is `&I` since it's passed to the IRBuilder constructor

sgundapa replied:
will look in to it

jmmartinez commented on llvm/test/Transforms/SeparateConstOffsetFromGEP/AMDGPU/xor-to-or-disjoint.ll:
```diff
@@ -0,0 +1,144 @@
+entry:
+  %base = add i64 2,0
+  %xor1 = xor i64 %base, 12
+  %xor2 = xor i64 24, %base    ; Operands reversed, should still be replaced
```
Shouldn't the ir cannonicalization passes ensure that the constant is on the right side? And other optimizations ensure that the rest of the constants are folded (including undefs) ?

sgundapa replied:
Yes. I have a piece of code that checks for constant in both positions. I wrote this test to validate that code. I will remove it and not be more pedantic w.r.t tests.

jmmartinez commented on llvm/lib/Transforms/Scalar/SeparateConstOffsetFromGEP.cpp:
```diff
@@ -1162,6 +1167,179 @@ bool SeparateConstOffsetFromGEP::splitGEP(GetElementPtrInst *GEP) {
+  bool FunctionChanged = false;
+  SmallVector<std::pair<Instruction *, Value *>, 16> ReplacementsToMake;
+
+  for (BasicBlock &BB : F) {
+    for (Instruction &I : BB) {
```
FYI in InstIterator there is a helper:
```suggestion
  for (Instruction &I : instructions(F)) {
```

sgundapa replied:
Will fix this. Thanks.

arsenm commented on llvm/lib/Transforms/Scalar/SeparateConstOffsetFromGEP.cpp:
```diff
@@ -1162,6 +1167,179 @@ bool SeparateConstOffsetFromGEP::splitGEP(GetElementPtrInst *GEP) {
+
+  LLVM_DEBUG(dbgs() << "SeparateConstOffsetFromGEP: Replacing" << I
+                    << " (used by GEP) with" << *NewOr << " based on"
+                    << *FoundUserInst << "\n");
```
```suggestion
                    << *FoundUserInst << '\n');
```

arsenm commented on llvm/lib/Transforms/Scalar/SeparateConstOffsetFromGEP.cpp:
```diff
@@ -1162,6 +1167,179 @@ bool SeparateConstOffsetFromGEP::splitGEP(GetElementPtrInst *GEP) {
+  BinaryOperator *UserBO = cast<BinaryOperator>(UserInst);
+  Value *UserOp0 = UserBO->getOperand(0);
+  Value *UserOp1 = UserBO->getOperand(1);
+  ConstantInt *UserC = nullptr;
+  if (UserOp0 == A)
+    UserC = dyn_cast<ConstantInt>(UserOp1);
+  else if (UserOp1 == A)
+    UserC = dyn_cast<ConstantInt>(UserOp0);
```
This doesn't handle the vector case, use PatternMatch to make it easier

sgundapa replied:
I will fix it.

arsenm commented on llvm/lib/Transforms/Scalar/SeparateConstOffsetFromGEP.cpp:
```diff
@@ -1162,6 +1167,179 @@ bool SeparateConstOffsetFromGEP::splitGEP(GetElementPtrInst *GEP) {
+Value *SeparateConstOffsetFromGEP::tryFoldXorToOrDisjoint(Instruction &I) {
+  assert(I.getOpcode() == Instruction::Xor && "Instruction must be XOR");
+
+  // Check if I has at least one GEP user.
+  bool HasGepUser = false;
+  for (User *U : I.users()) {
+    if (isa<GetElementPtrInst>(U)) {
+      HasGepUser = true;
+      break;
+    }
+  }
```
Generally shouldn't be looking at users but I also don't see why this cares

arsenm commented on llvm/lib/Transforms/Scalar/SeparateConstOffsetFromGEP.cpp:
```diff
@@ -1162,6 +1167,179 @@ bool SeparateConstOffsetFromGEP::splitGEP(GetElementPtrInst *GEP) {
+  return ClosestUser;
+}
+
+/// Try to transform I = xor(A, C1) into or_disjoint(Y, C2)
+/// where Y = xor(A, C0) is another existing instruction dominating I,
+/// C2 = C1 - C0, and A is known to be disjoint with C2.
```
This isn't core to what this pass is doing. This should have occurred in previous passes

jrbyrnes commented on llvm/lib/Transforms/Scalar/SeparateConstOffsetFromGEP.cpp:
```diff
@@ -1181,6 +1359,10 @@ bool SeparateConstOffsetFromGEP::run(Function &F) {
   bool Changed = false;
+
+  // Decompose xor in to "or disjoint" if possible.
+  Changed |= decomposeXor(F);
```
Is it possible to integrate this with the main function walk? I don't see why we need another walk.

sgundapa replied:
I will look at it.  I can change the main function walk, but it may it more complex. decomposeXor modifies the IR in such a way that the rest of the pass does its things.

arsenm commented:
> I tried doing this in InstCombine and the pass fail to reach a fixed point state and the pass is trying to merge back the two constants in to one. This is one of the reasons why I chose to do it SeparateConstOffset pass.
> 

Do you mean cases like the add 0, 0 from the tests? I'm confused about what this is trying to do . You should not be trying to create these. These also should not be in these tests (other than maybe one if you need to check a degenerate case doesn't crash)

sgundapa commented:
> Do you mean cases like the add 0, 0 from the tests?

I added this  (add 0,0) as a short cut for the knownbits to compute bits easily. I can rewrite them with variable, and, or to set the values i need.

sgundapa commented on llvm/lib/Transforms/Scalar/SeparateConstOffsetFromGEP.cpp:
```diff
@@ -1162,6 +1167,179 @@ bool SeparateConstOffsetFromGEP::splitGEP(GetElementPtrInst *GEP) {
+      if (I.getOpcode() == Instruction::Xor) {
+        if (Value *Replacement = tryFoldXorToOrDisjoint(I)) {
+          ReplacementsToMake.push_back({&I, Replacement});
+          FunctionChanged = true;
```
Sure.

sgundapa commented on llvm/lib/Transforms/Scalar/SeparateConstOffsetFromGEP.cpp:
```diff
@@ -1162,6 +1167,179 @@ bool SeparateConstOffsetFromGEP::splitGEP(GetElementPtrInst *GEP) {
+  Builder.SetInsertPoint(&I);
+  Constant *C2_Const = ConstantInt::get(Ty, C2_APInt);
+  Twine Name = I.getName();
+  Value *NewOr = BinaryOperator::CreateDisjointOr(FoundUserInst, C2_Const, Name,
```
Sure

sgundapa committed:
[SeparateConstOffsetFromGEP] Decompose constant xor operand if possible

Try to transform XOR(A, B+C) in to XOR(A,C) + B where XOR(A,C) becomes
the base for memory operations. This transformation is true under the
following conditions
Check 1 -  B and C are disjoint.
Check 2 - XOR(A,C) and B are disjoint.

This transformation is beneficial particularly for GEPs because
Disjoint OR operations often map better to addressing modes than XOR.
This can enable further optimizations in the GEP offset folding pipeline

sgundapa committed:
Removed dead code

jmmartinez commented on llvm/lib/Transforms/Scalar/SeparateConstOffsetFromGEP.cpp:
```diff
@@ -486,6 +490,39 @@ class SeparateConstOffsetFromGEP {
+  XorBaseValMap XorGroups;
+
+  /// Checks if the given value has at least one GetElementPtr user
+  bool hasGEPUser(const Value *V) const;
```
Function doesn't access the state of the object.
```suggestion
  static bool hasGEPUser(const Value *V);
```

jmmartinez commented on llvm/lib/Transforms/Scalar/SeparateConstOffsetFromGEP.cpp:
```diff
@@ -1162,6 +1199,165 @@ bool SeparateConstOffsetFromGEP::splitGEP(GetElementPtrInst *GEP) {
+
+  SmallVector<Instruction *, 8> InstructionsToErase;
+  const APInt SmallestConst =
+      dyn_cast<ConstantInt>(XorWithSmallConst->getOperand(1))->getValue();
```
```suggestion
      cast<ConstantInt>(XorWithSmallConst->getOperand(1))->getValue();
```

jmmartinez commented on llvm/lib/Transforms/Scalar/SeparateConstOffsetFromGEP.cpp:
```diff
@@ -1162,6 +1199,165 @@ bool SeparateConstOffsetFromGEP::splitGEP(GetElementPtrInst *GEP) {
+                 << "  New Const: " << NewConstVal << "\n");
+    }
+  }
+  if (!InstructionsToErase.empty())
+    for (Instruction *I : InstructionsToErase)
+      I->eraseFromParent();
```
```suggestion
  for (Instruction *I : InstructionsToErase)
    I->eraseFromParent();
```

jmmartinez commented on llvm/lib/Transforms/Scalar/SeparateConstOffsetFromGEP.cpp:
```diff
@@ -1162,6 +1199,165 @@ bool SeparateConstOffsetFromGEP::splitGEP(GetElementPtrInst *GEP) {
+      BinaryOperator *ClonedXor =
+          cast<BinaryOperator>(XorWithSmallConst->clone());
+      ClonedXor->setName(XorWithSmallConst->getName() + ".dom_clone");
+      ClonedXor->insertAfter(dyn_cast<Instruction>(OriginalBaseVal));
```
I assume that `dyn_cast<Instruction>(OriginalBaseVal)` would never be `nullptr`.

Otherwise, what if `OriginalBaseVal` is a global or an argument ?

If `OriginalBaseVal` can only be of type `Instruction`, we should enforce this in `OriginalBaseVal` type up to where the `OriginalBaseVal` is obtained.

sgundapa committed:
Address comments

jmmartinez commented on llvm/lib/Transforms/Scalar/SeparateConstOffsetFromGEP.cpp:
```diff
@@ -1162,6 +1199,167 @@ bool SeparateConstOffsetFromGEP::splitGEP(GetElementPtrInst *GEP) {
+
+    // Check if the XorWithSmallConst dominates currentXorToProcess.
+    // If not, clone and add the instruction.
+    if (!DT.dominates(XorWithSmallConst, currentXorToProcess)) {
```
Consider having a helper function to check if "the xor-with-small-const dominates all the other xor".

It's be easier to read

```cpp
if(xor-with-small-const doesn't dominate all other xors) {
  // do the clone
}
```

than

```cpp
for(xor in xors) {
  if(xor-with-small-const doesn't dominate xor) {
    // do the clone
    break;
  }
}
```

jmmartinez commented on llvm/lib/Transforms/Scalar/SeparateConstOffsetFromGEP.cpp:
```diff
@@ -1162,6 +1199,167 @@ bool SeparateConstOffsetFromGEP::splitGEP(GetElementPtrInst *GEP) {
+
+      auto *NewOrInst = BinaryOperator::CreateDisjointOr(
+          XorWithSmallConst,
+          ConstantInt::get(OriginalBaseInst->getType(), NewConstVal),
```
In some cases `NewConstVal` can be `0` (there is one in the tests). In that case you could skip creating the `disjoint or`.

sgundapa committed:
Address reviewer comments, take 2

jmmartinez reviewed (approved):
LGTM, but check my last comment to confirm we either handle unreachable basic-blocks / or that they are impossible.

jayfoad commented:
> Try to transform I = xor(A, C1) into or disjoint(Y, C2) where Y = xor(A, C0) is another existing instruction dominating I,
> C2 = C1 - C0, and A is known to be disjoint with C2.

This transform also requires C0 and C2 to be disjoint, right? So `C2 = C1 - C0` can also be written as `C2 = C1 ^ C0`.

jayfoad commented on llvm/lib/Transforms/Scalar/SeparateConstOffsetFromGEP.cpp:
```diff
@@ -1162,6 +1202,165 @@ bool SeparateConstOffsetFromGEP::splitGEP(GetElementPtrInst *GEP) {
 
+// Helper function to check if an instruction has at least one GEP user
+bool XorToOrDisjointTransformer::hasGEPUser(const Value *V) {
+  for (const User *U : V->users()) {
```
Use `any_of`?

arsenm commented on llvm/lib/Transforms/Scalar/SeparateConstOffsetFromGEP.cpp:
```diff
@@ -1162,6 +1202,165 @@ bool SeparateConstOffsetFromGEP::splitGEP(GetElementPtrInst *GEP) {
+    APInt NewConstVal = ConstOffsetVal - SmallestConst;
+    if ((NewConstVal & SmallestConst) != 0) {
+      LLVM_DEBUG(dbgs() << DEBUG_TYPE << ": Cannot transform XOR in function "
+                        << F.getName() << ":\n"
+                        << "  New Const: " << NewConstVal << "\n"
+                        << "  Smallest Const: " << SmallestConst << "\n"
+                        << "  are not disjoint \n");
```
```suggestion
                        << F.getName() << ":\n"
                        << "  New Const: " << NewConstVal 
                        << "  Smallest Const: " << SmallestConst << " are not disjoint\n");
```

arsenm commented on llvm/test/Transforms/SeparateConstOffsetFromGEP/AMDGPU/xor-to-or-disjoint.ll:
```diff
@@ -0,0 +1,154 @@
+  %val1 = load half, ptr %gep1
+  %val2 = load half, ptr %gep2
+  %val3 = load half, ptr %gep3
+  ret void
```
Fix this to not be dead code

arsenm reviewed (changes requested):
InstCombine already removes these xors. Your examples have them all fold away when run through instcombine. In general we should not need to worry about non-canonical IR in late optimization passes

arsenm commented:
> I tried doing this in InstCombine and the pass fail to reach a fixed point state and the pass is trying to merge back the two constants in to one.

It would be better to review this attempt


 This is one of the reasons why I chose to do it SeparateConstOffset pass.
> 
> Sample IR code:
> 
> %155 = or disjoint i32 %153, %154 %156 = getelementptr inbounds nuw half, ptr addrspace(3) @global_smem, i32 %155 %157 = xor i32 %155, 32 %158 = getelementptr inbounds nuw half, ptr addrspace(3) @global_smem, i32 %157 %159 = or disjoint i32 %155, 2048 %160 = getelementptr inbounds nuw half, ptr addrspace(3) @global_smem, i32 %159 %161 = xor i32 %155, 2080 %162 = getelementptr inbounds nuw half, ptr addrspace(3) @global_smem, i32 %161 %163 = or disjoint i32 %155, 4096 %164 = getelementptr inbounds nuw half, ptr addrspace(3) @global_smem, i32 %163 %165 = xor i32 %155, 4128 %166 = getelementptr inbounds nuw half, ptr addrspace(3) @global_smem, i32 %165 %167 = or disjoint i32 %155, 6144 %168 = getelementptr inbounds nuw half, ptr addrspace(3) @global_smem, i32 %167 %169 = xor i32 %155, 6176 %170 = getelementptr inbounds nuw half, ptr addrspace(3) @global_smem, i32 %169 %171 = or disjoint i32 %155, 8192 %172 = getelementptr inbounds nuw half, ptr addrspace(3) @global_smem, i32 %171 %173 = xor i32 %155, 8224 %174 = getelementptr inbounds nuw half, ptr addrspace(3) @global_smem, i32 %173


Samples that are compilable are always more helpful, this is just a fragment I can't run experiments with

sgundapa commented:
> > Try to transform I = xor(A, C1) into or disjoint(Y, C2) where Y = xor(A, C0) is another existing instruction dominating I,
> > C2 = C1 - C0, and A is known to be disjoint with C2.
> 
> This transform also requires C0 and C2 to be disjoint, right? So `C2 = C1 - C0` can also be written as `C2 = C1 ^ C0`.

I have updated the description in a subsequent commit message. This is what I ended up with 
"Try to transform XOR(A, B+C) in to XOR(A,C) + B where XOR(A,C) becomes
the base for memory operations. This transformation is true under the
following conditions
Check 1 -  B and C are disjoint.
Check 2 - XOR(A,C) and B are disjoint"

I can check for A and B are disjointed as well, but check 1 and check 2  can deduce the same.

sgundapa committed:
Address reviewer comments , take 3

sgundapa commented:
I will address if the there any concerns, post the merge.

arsenm commented:
I think this should be reverted and abandoned. This has nothing to do with SeparateConstOffsetFromGEP, and is another patch bolted at the front of it. The provided tests are non-canonical IR and I suspect this is just working around the depth limit issue in instcombine that @jrbyrnes was working on

arsenm commented:
I have reverted this in ad479ddb343c2756e6eed0f2999bbdb88a65c7c5

sgundapa commented:
Increased recursion depth was needed to prove the disjointedness. Jeff has been involved to work on the issues related to recursion depth.  For the Triton use case w.r.t Linear Layout, the addressing mode they chose to represent lane/warp/block/reg offset involves xors. 
The unit tests are written to test the pass. I can write an unit test that can run both InstCombine and SeparateConstOffset pass, if needed.

shiltian commented:
> Ping! This is a P1 in downstream ticketing system.

This PR has NEVER been approved so you can't just merge it, even it is blocking anything, needless to say you basically already got request for change from two folks.
